xmtp_mls/src/groups/subscriptions.rs

```diff
@@ -44,35 +47,35 @@ where
     pub async fn stream(
         &'c self,
     ) -> Result<Pin<Box<dyn Stream<Item = StoredGroupMessage> + 'c + Send>>, GroupError> {
-        let last_cursor = 0;
-
-        let subscription = self
+        Ok(self
             .client
-            .api_client
-            .subscribe_group_messages(vec![GroupFilter::new(
+            .stream_messages(HashMap::from([(
                 self.group_id.clone(),
-                Some(last_cursor as u64),
-            )])
-            .await?;
-        let stream = subscription
-            .map(|res| async {
-                match res {
-                    Ok(envelope) => self.process_stream_entry(envelope).await,
-                    Err(err) => Err(GroupError::Api(err)),
-                }
-            })
-            .filter_map(move |res| async {
-                match res.await {
-                    Ok(Some(message)) => Some(message),
-                    Ok(None) => None,
-                    Err(err) => {
-                        log::error!("Error processing stream entry: {:?}", err);
-                        None
-                    }
-                }
-            });
-
-        Ok(Box::pin(stream))
+                MessagesStreamInfo {
+                    convo_created_at_ns: self.created_at_ns,
+                    cursor: 0,
+                },
+            )]))
+            .await?)
+    }
+
+    pub async fn stream_with_callback(
+        client: Arc<Client<ApiClient>>,
+        group_id: Vec<u8>,
+        created_at_ns: i64,
+        mut callback: impl FnMut(StoredGroupMessage) + Send + 'static,
```

Comment:
These callbacks are meant to be short-lived and discarded when the stream closes. Won't the `static tag actually keep them around forever?

Reply:
The same reasoning as above applies here. My understanding:

1. Adding `'static` to a declaration doesn't functionally change how long it lives for. It is just a compile-time check that whatever is passed in will actually live that long.
2. All structs are `'static` unless they hold a non-static reference
3. I'm adding `'static` to an implementer of the `FnMut` trait here, which is a closure that is struct-like (passed by value, not by reference). What I'm doing is enforcing that the closure that an external caller provides here doesn't hold any non-static references.

I hope this understanding is correct, would love it if @tuddman or @insipx could double-check this!

Reply:
`'static` by definition is guaranteed to live for as long as the lifetime of the program is resident in memory.  If something is passed in, it will remain there, forever.  
Point 4 is the unfortunately important one here.  Crossing thread boundaries relinquishes borrowing control.  The alt is to wrap it in an `Arc<T>` which exists, in part,  for this very reason.

Reply:
The main point I'm trying to convey here is I don't think this is sacrificing anything. By default, an `FnMut` passed to a foreign thread is already `'static`, because it is owned by the foreign thread and guaranteed to live as long as the thread does. The only way in which it would not be static is if the closure held onto non-static references from the surrounding environment. We're simply prohibiting that - anything the closure holds onto either has to be moved, wrapped in an Arc, or `'static` to begin with.

Reply:
Here's a thread explaining my current understanding: https://users.rust-lang.org/t/static-lifetime-with-owned-object/50637/4

> Lifetimes don't apply to types that don't borrow anything. This makes self-contained types appear to satisfy any lifetime, but in fact they ignore all lifetimes.
>
> T: 'static will allow a type like String. It's not because String lives for as long as 'static (it could be dropped at any time), but because it's not borrowing from any scope, so there's no lifetime limiting its usage.

Reply:
This is very useful context. Maybe my original comment is based on a poor understanding of how lifetimes apply to things that don't borrow.